File: src/Dota2Analyzer.Api/Services/ItemImageCache.cs
using Dota2Analyzer.Core.Services;

namespace Dota2Analyzer.Api.Services;

public sealed class ItemImageCache
{
    private readonly HeroDataCache _heroData;
    private readonly HttpClient _httpClient;
    private readonly string _itemImageDir;
    private readonly ILogger<ItemImageCache> _logger;
    private readonly string[] _baseUrls;

    public ItemImageCache(HeroDataCache heroData, ILogger<ItemImageCache> logger, IConfiguration config)
    {
        _heroData = heroData;
        _logger = logger;
        _httpClient = new HttpClient();
        _baseUrls = BuildBaseUrls(config);

        var basePath = Environment.GetFolderPath(Environment.SpecialFolder.LocalApplicationData);
        var folder = Path.Combine(basePath, "Dota2Analyzer", "assets", "items");
        Directory.CreateDirectory(folder);
        _itemImageDir = folder;
    }

    public async Task<string?> EnsureItemImageAsync(string itemKey, CancellationToken cancellationToken)
    {
        if (string.IsNullOrWhiteSpace(itemKey)) return null;

        var normalized = NormalizeKey(itemKey);
        if (string.IsNullOrWhiteSpace(normalized)) return null;

        var localPath = Path.Combine(_itemImageDir, $"{normalized}.png");
        if (File.Exists(localPath))
        {
            _logger.LogDebug("Item image cache hit {Path}", localPath);
            return localPath;
        }

        await _heroData.EnsureLoadedAsync(cancellationToken);
        if (!_heroData.TryGetItemConstants(normalized, out var constants) || constants is null)
        {
            _logger.LogWarning("Item constants missing for {Key}", normalized);
            return null;
        }

        if (string.IsNullOrWhiteSpace(constants.Img))
        {
            _logger.LogWarning("Item image path missing for {Key}", normalized);
            return null;
        }

        foreach (var url in BuildCandidateUrls(constants.Img))
        {
            _logger.LogInformation("Downloading item image {Url}", url);
            try
            {
                using var response = await _httpClient.GetAsync(url, cancellationToken);
                if (!response.IsSuccessStatusCode)
                {
                    _logger.LogWarning("Item image download failed {StatusCode} for {Url}", response.StatusCode, url);
                    continue;
                }

                await using var stream = await response.Content.ReadAsStreamAsync(cancellationToken);
                await using var fs = File.Create(localPath);
                await stream.CopyToAsync(fs, cancellationToken);
                _logger.LogInformation("Item image saved {Path}", localPath);
                return localPath;
            }
            catch (HttpRequestException ex)
            {
                _logger.LogWarning(ex, "Item image download failed for {Url}", url);
            }
        }

        return null;
    }

    private IEnumerable<string> BuildCandidateUrls(string imgPath)
    {
        if (string.IsNullOrWhiteSpace(imgPath))
        {
            yield break;
        }

        if (imgPath.StartsWith("http", StringComparison.OrdinalIgnoreCase))
        {
            yield return imgPath;
            yield break;
        }

        var trimmed = imgPath.StartsWith("/") ? imgPath : "/" + imgPath;
        foreach (var baseUrl in _baseUrls)
        {
            yield return $"{baseUrl.TrimEnd('/')}{trimmed}";
        }
    }

    private static string NormalizeKey(string key)
    {
        var normalized = key.Trim().ToLowerInvariant();
        if (normalized.StartsWith("item_"))
        {
            normalized = normalized[5..];
        }
        return normalized;
    }

    private static string[] BuildBaseUrls(IConfiguration config)
    {
        var primary = config["Assets:ItemImageBaseUrl"];
        var fallback = config["Assets:ItemImageFallbackBaseUrl"];

        var list = new List<string>();
        if (!string.IsNullOrWhiteSpace(primary)) list.Add(primary);
        if (!string.IsNullOrWhiteSpace(fallback)) list.Add(fallback);

        if (list.Count == 0)
        {
            list.Add("https://cdn.opendota.com");
            list.Add("https://cdn.cloudflare.steamstatic.com");
        }

        return list.Distinct().ToArray();
    }
}

File: dota2-analyzer-api/src/main/java/com/dota2analyzer/api/service/ItemImageCache.java
package com.dota2analyzer.api.service;

import com.dota2analyzer.core.model.opendota.ItemConstants;
import com.dota2analyzer.core.service.HeroDataCache;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import java.io.*;
import java.net.URI;
import java.net.http.HttpClient;
import java.net.http.HttpRequest;
import java.net.http.HttpResponse;
import java.nio.file.Files;
import java.nio.file.Path;
import java.nio.file.Paths;
import java.time.Duration;

public class ItemImageCache {

    private static final Logger log = LoggerFactory.getLogger(ItemImageCache.class);
    private static final String[] BASE_URLS = {
            "https://cdn.opendota.com",
            "https://cdn.cloudflare.steamstatic.com"
    };

    private final HeroDataCache heroData;
    private final HttpClient httpClient;
    private final Path itemImageDir;

    public ItemImageCache(HeroDataCache heroData) {
        this.heroData = heroData;
        this.httpClient = HttpClient.newBuilder().connectTimeout(Duration.ofSeconds(10)).build();

        String home = System.getProperty("user.home");
        this.itemImageDir = Paths.get(home, ".dota2analyzer", "assets", "items");
        try {
            Files.createDirectories(itemImageDir);
        } catch (IOException e) {
            log.warn("Failed to create item image directory", e);
        }
    }

    public String ensureItemImage(String itemKey) {
        if (itemKey == null || itemKey.isBlank()) return null;

        String normalized = normalizeKey(itemKey);
        if (normalized.isBlank()) return null;

        Path localPath = itemImageDir.resolve(normalized + ".png");
        if (Files.exists(localPath)) {
            log.debug("Item image cache hit {}", localPath);
            return localPath.toString();
        }

        heroData.ensureLoaded();
        ItemConstants constants = heroData.getItemConstants(normalized);
        if (constants == null || constants.getImg() == null || constants.getImg().isBlank()) {
            log.warn("Item constants or image path missing for {}", normalized);
            return null;
        }

        String imgPath = constants.getImg();

        // If absolute URL
        if (imgPath.startsWith("http")) {
            return downloadImage(imgPath, localPath);
        }

        // Try with base URLs
        String trimmed = imgPath.startsWith("/") ? imgPath : "/" + imgPath;
        for (String baseUrl : BASE_URLS) {
            String result = downloadImage(baseUrl + trimmed, localPath);
            if (result != null) return result;
        }

        return null;
    }

    private String downloadImage(String url, Path localPath) {
        log.info("Downloading item image {}", url);
        try {
            HttpRequest request = HttpRequest.newBuilder()
                    .uri(URI.create(url))
                    .timeout(Duration.ofSeconds(10))
                    .GET().build();
            HttpResponse<InputStream> response = httpClient.send(request, HttpResponse.BodyHandlers.ofInputStream());
            if (response.statusCode() != 200) {
                log.warn("Item image download failed {} for {}", response.statusCode(), url);
                return null;
            }
            try (InputStream is = response.body();
                 OutputStream os = new BufferedOutputStream(Files.newOutputStream(localPath))) {
                is.transferTo(os);
            }
            log.info("Item image saved {}", localPath);
            return localPath.toString();
        } catch (Exception ex) {
            log.warn("Item image download failed for {}", url, ex);
            return null;
        }
    }

    private static String normalizeKey(String key) {
        String normalized = key.trim().toLowerCase();
        if (normalized.startsWith("item_")) {
            normalized = normalized.substring(5);
        }
        return normalized;
    }
}
